cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,371 @@
+
+## Use flat Promise chains
+
+Asynchronous callback functions are one of the strongest features of Node.js. However, increasing layers of nesting within callback functions can become a problem. Any multistage process can become nested 10 or more levels deep. This problem is called as Pyramid of Doom or Callback Hell. In such a code, the errors and results get lost within the callback. Promises are a good way to write asynchronous code without getting into nested pyramids. Promises provide top-down execution while being asynchronous by delivering errors and results to next .then function. Another advantage of Promises is the way Promises handle the errors. If an error occurs in a Promise class, it skips over the .then functions and invokes the first .catch function it finds. This way Promises bring a higher assurance of capturing and handling errors. As a principle, you can make all your asynchronous code(apart from emitters) return promises. However, it should be noted that Promise calls can also become a pyramid. In order to completely stay away from callback hells, flat Promise chains should be used. If the module you are using does not support Promises, you can convert base object to a Promise by using Promise.promisifyAll() function.
+
+## Return sanitized user objects
+
+Information about the users of an application is among the most critical information about the application. Therefore, user information should be stored on a different database. User tables generally include fields like id, username, full name, email address, birth date, credit card number and in some cases social security numbers. Therefore, when querying and using user objects, you need to return only needed fields as it may be vulnerable to personal information disclosure. This is also correct for other objects stored on the database. If you just need a certain field of an object, you should never return it with all of its fields. As an example, you can use a function like the following whenever you need to get information on a user. By doing so, you can only return the fields that are needed for your specific operation. In other words, if you only need to list names of the users available, you are not returning their email addresses or credit card numbers in addition to their full names.
```

Comment:
Not sure what you mean by `stored on a different database`?

`credit card number should never be stored in a users table. Maybe `password` instead?

`you should only return the specific fields required` rather than `you should never return it with all the fields`

Reply:
changed accordingly